keras/src/dtype_policies/dtype_policy_map.py

```diff
@@ -0,0 +1,215 @@
+import re
+from collections.abc import MutableMapping
+
+from keras.src import dtype_policies
+from keras.src.api_export import keras_export
+from keras.src.dtype_policies import DTypePolicy
+
+
+@keras_export(["keras.dtype_policies.DTypePolicyMap"])
+class DTypePolicyMap(DTypePolicy, MutableMapping):
+    """A dict-like object that maps string to `DTypePolicy` instances.
+
+    `DTypePolicyMap` can be used in `get_config` in layers and subclasses to
+    support a complex configurations of dtype policies.
+
+    For example, we can modify `get_config` in `layers.MultiHeadAttention` as
+    follows to support the mixing of dtype policies, such as quantization.
+
+    ```python
+    @keras.saving.register_keras_serializable("MyPackage")
+    class MyMultiHeadAttention(keras.layers.MultiHeadAttention):
+        def get_config(self):
+            config = super().get_config()
+            dtype_policy_map = dtype_policies.DTypePolicyMap()
+            for layer in self._flatten_layers():
+                if layer.dtype_policy.is_quantized:
+                    dtype_policy_map[layer.name] = layer.dtype_policy
+            if len(dtype_policy_map) > 0:
+                config.update({"dtype": dtype_policy_map})
+            return config
+    ```
+
+    Internally, `DTypePolicyMap` uses a string as key and a `DTypePolicy`
+    as value. There is a behavior difference between a normal Python dict and
+    this class. The string key will be treated as a regex when retrieving the
+    value. See the docstring of `get` for more details.
+
+    See below for a usage example. You can define the naming schema
+    of the `DTypePolicy`, and then retrieve the corresponding `DTypePolicy`
+    instance.
+
+    In the normal case, the key to query is usually the `layer.name`, which
+    is the `name` of the layer.
+
+    ```python
+    dtype_policy_map = DTypePolicyMap()
+    dtype_policy_map["layer/dense_0"] = FloatDTypePolicy("bfloat16")
```

Comment:
Should we introduce a `layer.path` property similar to `variable.path`?

Reply:
Introducing `Layer.path` makes it much cleaner. Done.